### PR TITLE
Add binary ops for _foreach APIs

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43926 [Wip] Rewrote Adadelta with _foreach APIs
* #43925 [wip] Rewrote adamax optimizer with _foreach APIs
* #43727 WIP asgd
* #43726 [wip] Rewrote rprop optimizer with _foreach APIs
* #43725 [wip] Rewrote rmsprop with _forerach APIs
* #43724 [wip] Rewrote SGD with _foreach APIs
* #43562 [wip] Rewrote AdamW with _foreach APIs
* #43507 [WIP] Rewrote adam optimizer with foreach APIs
* #43413 Added alpha overloads for add/sub ops with lists
* #42537 Add unary ops: exp, sqrt and pointwise ops: addcdiv, addcmul
* **#42536 Add binary ops for _foreach APIs**

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
Adding APIs:
```
torch._foreach_sub(TensorList tl1, TensorList tl2)
torch._foreach_sub_(TensorList self, TensorList tl2)
torch._foreach_mul(TensorList tl1, TensorList tl2)
torch._foreach_mul_(TensorList self, TensorList tl2)
torch._foreach_div(TensorList tl1, TensorList tl2)
torch._foreach_div_(TensorList self, TensorList tl2)

torch._foreach_sub(TensorList tl1, Scalar scalar)
torch._foreach_sub_(TensorList self, Scalar scalar)
torch._foreach_mul(TensorList tl1, Scalar scalar)
torch._foreach_mul_(TensorList self, Scalar scalar)
torch._foreach_div(TensorList tl1, Scalar scalar)
torch._foreach_div(TensorList self, Scalar scalar)
```

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists
2. Properly handle bool tensors

**Plan for the next PRs**
1. APIs
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331891](https://our.internmc.facebook.com/intern/diff/D23331891)

### CODE REVIEW

File: aten/src/ATen/native/ForeachOpsKernels.cpp [hunk 1]
```diff
@@ -42,4 +98,61 @@ void foreach_tensor_add_list_kernel_slow_(TensorList tensors1, TensorList tensor
...
+
+void foreach_tensor_mul_list_kernel_slow_(TensorList tensors1, TensorList tensors2) {
+  verify_list(tensors1, tensors2);
+
+  for (int i = 0; i < tensors1.size(); i++) {
+    tensors1[i].mul_(tensors2[i]);
+  }
+}
+
+std::vector<Tensor> foreach_tensor_div_list_kernel_slow(TensorList tensors1, TensorList tensors2) {
```
Comment (zou3519) [inline aten/src/ATen/native/ForeachOpsKernels.cpp:139]: Would prefer some templates here for DRY. e.g.,

```
template <Tensor (*Op)(const Tensor&, const Tensor&)>
void foreach_tensor_op_list_kernel_slow_(TensorList tensors1, TensorList tensors2) {
    verify_list(tensors1, tensors2)
    for (...) {
        result.emplace_back(Op(tensors1[i], tensors2[i]);
    }
}
```

File: aten/src/ATen/native/cuda/ForeachBinaryOpList.cu [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+std::vector<Tensor> foreach_tensor_div_list_kernel_cuda(TensorList tensors1, TensorList tensors2) {
+    verify_list(tensors1, tensors2);
+
+    if (!check_fast_route(tensors1, tensors2)) {
+        return at::native::foreach_tensor_div_list_kernel_slow(tensors1, tensors2);
+    }
+
+    return foreach_tensor_list_op<std::divides>(tensors1, tensors2);
+}
```
Comment (zou3519) [inline aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:105]: We could toss the slow route checks into a template as well if we had something like:
```
template<Tensor (*AtOp)(const Tensor&, const Tensor&), template<class> class Op>
std::vector<Tensor> foreach_tensor_list_op(TensorList tensors1, TensorList tensors2) {
    verify_list(tensors1, tensors2);
    if (!check_fast_route(tensors1, tensors2) {
        native::foreach_tensor_list_kernel_slow<AtOp>(tensors1, tensors2);
    }
    std::vector<std::vector<at::Tensor>> tensor_lists; 
    std::vector<at::Tensor> vec_res;
    for (const auto& t: tensors1) {
        vec_res.emplace_back(at::native::empty_like(t));
    }

    tensor_lists.emplace_back(std::move(tensors1.vec()));
    tensor_lists.emplace_back(std::move(tensors2.vec()));
    tensor_lists.emplace_back(std::move(vec_res));

    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, tensors1[0].scalar_type(), "foreach_binary_op_list_cuda", [&]() {
        multi_tensor_apply<3>(tensor_lists, BinaryOpListFunctor<scalar_t, Op>());
    });

    return tensor_lists[2];
}

std::vector<Tensor> foreach_tensor_div_list_kernel_cuda(TensorList tensors1, TensorList tensors2) {
  return foreach_tensor_list_op<at::div, std::divides>(tensors1, tensors2);
}
```

File: aten/src/ATen/native/ForeachOpsKernels.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/ForeachOpsKernels.cpp:139]: Actually, I see the conundrum. native_functions.yaml wants the name of the function and probably can't take a template.

Comment (cpuhrsch) [reply aten/src/ATen/native/ForeachOpsKernels.cpp:139]: We could use function pointers that point to an instantiated version of the above template. But all of that seems like overkill for now. Why not use a simple macro?

File: test/test_foreach.py [hunk 3]
```diff
@@ -173,14 +184,71 @@ def test_add_list_error_cases(self, device):
...
 
+        # add
         res = torch._foreach_add(tensors1, tensors2)
         torch._foreach_add_(tensors1, tensors2)
         self.assertEqual(res, tensors1)
-        self.assertEqual(res[0], torch.ones(10, 10, device=device, dtype=dtype))
+        self.assertEqual(tensors1, [torch.ones(20, 20, device=device, dtype=dtype) for _ in range(20)])
+
+        # sub
+        res = torch._foreach_sub(tensors1, tensors2)
```
Comment (cpuhrsch) [inline test/test_foreach.py:201]: You could parameterize this function like

```
def _test(fn):
 ...
_test(torch._foreach_mul, torch.mul)
```
and then you have to write this code only once.

File: test/test_foreach.py [hunk 4]
```diff
@@ -173,14 +184,71 @@ def test_add_list_error_cases(self, device):
...
 
     @dtypes(*torch.testing.get_all_dtypes())
-    def test_add_list_same_size(self, device, dtype):
-        tensors1 = [torch.zeros(10, 10, device=device, dtype=dtype) for _ in range(10)]
-        tensors2 = [torch.ones(10, 10, device=device, dtype=dtype) for _ in range(10)]
+    def test_bin_op_list(self, device, dtype):
+        if dtype == torch.bool:
+            return
+
+        tensors1 = [torch.zeros(20, 20, device=device, dtype=dtype) for _ in range(20)]
```
Comment (cpuhrsch) [inline test/test_foreach.py:191]: Why is all of this zeros?

File: test/test_foreach.py [hunk 5]
```diff
@@ -173,14 +184,71 @@ def test_add_list_error_cases(self, device):
             torch._foreach_add_(tensors1, tensors2)
 
     @dtypes(*torch.testing.get_all_dtypes())
-    def test_add_list_same_size(self, device, dtype):
-        tensors1 = [torch.zeros(10, 10, device=device, dtype=dtype) for _ in range(10)]
-        tensors2 = [torch.ones(10, 10, device=device, dtype=dtype) for _ in range(10)]
+    def test_bin_op_list(self, device, dtype):
+        if dtype == torch.bool:
+            return
```
Comment (ngimel) [inline test/test_foreach.py:189]: What happens with bool? Should you be checking for runtime error with bool instead of bailing out completely?

File: test/test_foreach.py (hunk 3 above)
Comment (izdeby) [reply test/test_foreach.py:201]: The reason for ones and zeros is that we are testing all the dtypes and there is no random method that supports all. So im using ones and zeros to test that all ops are working as expected

File: test/test_foreach.py (hunk 4 above)
Comment (izdeby) [reply test/test_foreach.py:191]: see below, please

File: test/test_foreach.py (hunk 3 above)
Comment (ngimel) [reply test/test_foreach.py:201]: You still can create a helper function that produces a tensor of requested dtype. It could check dtype and depending on the dtype use either randint or rand. 
Right now the test flow is confusing, e.g. here you rely on `tensors1` being set to 0 by an inplace add test, but `add` and `sub` tests really should be independent, and parametrizing them as @cpuhrsch suggests would allow you to do that.

File: aten/src/ATen/native/cuda/ForeachBinaryOpList.cu [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+template<template<class> class Op>
+std::vector<Tensor> foreach_tensor_list_op(TensorList tensors1, TensorList tensors2) {
+    std::vector<std::vector<at::Tensor>> tensor_lists; 
+    std::vector<at::Tensor> vec_res;
+    for (const auto& t: tensors1) {
+        vec_res.emplace_back(at::native::empty_like(t));
+    }
+
+    tensor_lists.emplace_back(std::move(tensors1.vec()));
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:15]: This ".vec" operation is very expensive last time I checked. Might be worth doing a for-loop instead of calling into .vec.

File: aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, tensors[0].scalar_type(), "foreach_binary_op_scalar_cuda_", [&]() {
+        multi_tensor_apply<1>(tensor_lists, BinaryOpScalarFunctor_<scalar_t, Op>(), scalar.to<scalar_t>());
+    });
+}
+
+#define FOREACH_BINARY_OP_SCALAR(NAME, OP)                                                          \
+void foreach_tensor_##NAME##_scalar_kernel_cuda_(TensorList tensors, Scalar scalar) {               \
+    check_foreach_api_restrictions(tensors);                                                        \
+                                                                                                    \
+    if (!can_use_fast_route(tensors, scalar)) {                                                     \
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu:42]: nit: I personally find it clearer to write

```
if (can_use_fast_route(tensors, scalar)) { 
     foreach_binary_op_<OP>(tensors, scalar);
    return;
}
at::native::foreach_tensor_##NAME##_scalar_kernel_slow_(tensors, scalar);
}
```

to avoid that extract negative.

File: test/test_foreach.py [hunk 8]
```diff
@@ -4,6 +4,17 @@
...
+    bin_ops = [
+        torch._foreach_add, 
+        torch._foreach_add_,
+        torch._foreach_sub,
+        torch._foreach_sub_,
+        torch._foreach_mul, 
+        torch._foreach_mul_, 
+        torch._foreach_div, 
+        torch._foreach_div_,
+        ]
```
Comment (cpuhrsch) [inline test/test_foreach.py:16]: seems like a legit lint failure

File: aten/src/ATen/native/cuda/ForeachBinaryOpList.cu (hunk 6 above)
Comment (zou3519) [reply aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:15]: The expensive part of .vec() is the dynamic allocation involved in creating a new vector. tensors1.vec() creates a new vector. Doing a for-loop to push everything in tensors1 to a new vector also creates a new vector, so they should be the same in terms of performance, right?

File: test/test_foreach.py (hunk 5 above)
Comment (izdeby) [reply test/test_foreach.py]: Fixed

File: test/test_foreach.py (hunk 3 above)
Comment (izdeby) [reply test/test_foreach.py]: Done

**Comment by mcarilli** [comment]
There are in-place and out-of-place variants, which is good.  But something's missing,  an `out=Tensor[] out` variant.

an `out=` variant is especially powerful with MTA because it lets you out create and pass an `out=` list as sliced views of a flat buffer.

More generally, I think the available variants should mimic what's currently present for single-tensor binary ops.  For example, here's what currently exists for `sub` in native_functions.yaml:
```
- func: sub.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
  dispatch:
    CPU, CUDA: sub_out
    SparseCPU, SparseCUDA: sub_out_sparse

- func: sub.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
  use_c10_dispatcher: full
  variants: function, method
  dispatch:
    CPU, CUDA: sub
    SparseCPU, SparseCUDA: sub_sparse

- func: sub_.Tensor(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)
  use_c10_dispatcher: full
  variants: method
  dispatch:
    CPU, CUDA: sub_
    SparseCPU, SparseCUDA: sub_sparse_

# For C++ only, until we have conversion from C++ numbers to Tensor
- func: sub.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> Tensor
  use_c10_dispatcher: full
  variants: function, method

- func: sub_.Scalar(Tensor(a!) self, Scalar other, Scalar alpha=1) -> Tensor(a!)
  use_c10_dispatcher: full
  variants: method
```

I think the most consistent thing to do is mimic that interface for all our binary ops (_including_ the `alpha` arg, which isn't difficult to add to the implementation or functor signatures).

**Comment by izdeby** [comment]
> There are in-place and out-of-place variants, which is good. But something's missing, an `out=Tensor[]` variant.
> 
> an `out=` variant is especially powerful with MTA because it lets you out create and pass an `out=` list as sliced views of a flat buffer, so from Python or C++ client code it becomes simple to effectively fuse a cat with a binary op. (Using an MTA `div_(grads_this_bucket, world_size, out=views_sliced_from_bucket)` is my plan for gradient bucketing in Reducer: an entire bucket's worth gradients can be avged by world size and copied into the flat bucket with a single kernel launch. Right now I fuse div and copy using mul_out, but that launches one kernel per grad, so it incurs more CPU overhead.)
> 
> More generally, I think the available variants should mimic what's currently present for single-tensor binary ops. For example, here's what currently exists for `sub` in native_functions.yaml:
> 
> ```
> - func: sub.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
>   dispatch:
>     CPU, CUDA: sub_out
>     SparseCPU, SparseCUDA: sub_out_sparse
> 
> - func: sub.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
>   use_c10_dispatcher: full
>   variants: function, method
>   dispatch:
>     CPU, CUDA: sub
>     SparseCPU, SparseCUDA: sub_sparse
> 
> - func: sub_.Tensor(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)
>   use_c10_dispatcher: full
>   variants: method
>   dispatch:
>     CPU, CUDA: sub_
>     SparseCPU, SparseCUDA: sub_sparse_
> 
> # For C++ only, until we have conversion from C++ numbers to Tensor
> - func: sub.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> Tensor
>   use_c10_dispatcher: full
>   variants: function, method
> 
> - func: sub_.Scalar(Tensor(a!) self, Scalar other, Scalar alpha=1) -> Tensor(a!)
>   use_c10_dispatcher: full
>   variants: method
> ```
> 
> I think the most consistent thing to do is have MTA interfaces imitate single-tensor interfaces for all binary ops (_including_ the `alpha` arg, which isn't difficult to add to the MTA implementation or OpFunctor signatures).

As discussed, i agree that we need this and adding this to the TODO section.

**Review by cpuhrsch — APPROVED** [review]
LGTM
